WIP BXMSDOC-8000 & BXMSDOC-8001: Update Kogito book as per 7.12 STP BAPLs (#72)
hmanwani-rh/kie-docs: doc-content/enterprise-only/getting-started-kogito/proc-kogito-remote-process-decision-integrating.adoc

[id="proc-kogito-remote-process-decision-integrating_{context}"]
= Integrating process and decision microservices in {KOGITO} using the remote method

[role="_abstract"]
The remote method uses REST endpoints from the decision microservices to integrate the process and decision microservices. In the REST endpoints, the process microservice can use BPMN components to send HTTP requests, such as a service task or a REST work item.

> The remote method does not require the process and decision microservices to run in the same application, since the communication is performed using the network.

Reviewer: Change since to because and search for since in the doc.

Author: Done

.Remote method of integrating process and decision microservices
image::KogitoMicroservices/process-decisions-rest-integration.png[Image of Process integration with decisions through REST]

> You can define the decisions in different domains or assets, such as DMN and DRL. In the REST endpoints, the decision microservices are deployed from the process microservice. Also, the REST endpoints are generated under the same application for the process and decision microservices, but using different methods. However, the endpoints can be placed in different applications.

Reviewer 2: ```suggestion
You can define the decisions in different domains or assets, such as DMN and DRL. In the remote method, the decision microservices are deployed from the process microservice. Also, the REST endpoints are generated under the same application for the process and decision microservices. However, the endpoints can be placed in different applications.
```
REST endpoints methods are the same, just endpoint names are different. So endpoints in this example are part of one application but in general it may be separated. I think this is the main message, isn't it?

Author: With the phrase `but using different methods` it means the REST endpoints for processes and decision microservices are generated using different methods, however under the same application. 
Do you think we can keep that phrase?

Reviewer 2: Yeah, we can keep `using different methods`, but I would change `REST endpoints` to `remote method` anyway.

> You can also use the `process-decisions-rest-quarkus` and `process-decisions-rest-springboot` example applications for {QUARKUS} and Spring Boot, respectively. When an example application is running, there also must be different process microservices that are running, such as:

Reviewer: ```suggestion
You can also use the `process-decisions-rest-quarkus` and `process-decisions-rest-springboot` example applications for {QUARKUS} and Spring Boot, respectively. When an example application is running, there must also be different process microservices that are running, such as:
```

Author: Done

> You can also use the `process-decisions-rest-quarkus` and `process-decisions-rest-springboot` example applications for {QUARKUS} and Spring Boot, respectively. When an example application is running, there must also be different process microservices that are running, such as:

Reviewer 2: ```suggestion
You can also use the `process-decisions-rest-quarkus` and `process-decisions-rest-springboot` example applications for {QUARKUS} and Spring Boot, respectively. When an example application is running, there are two different process microservices that are running, such as:
```

Author: Changed it to `You can also use the `process-decisions-rest-quarkus` and `process-decisions-rest-springboot` example applications for {QUARKUS} and Spring Boot, respectively. When an example application is running, two different processes microservices also run, such as:`

----

.Example response
[source,json]
----
{
  "id": "e499326d-3bd2-4ddb-93b7-0f68f74a9673",
  "driverId": "12345",
  "driver": {
    "licenseExpiration": "2021-08-12T19:35:48.971+00:00",
    "validLicense": true,
    "Name": "Arthur",
    "State": "SP",
    "City": "Campinas",
    "Points": 13,
    "Age": 30
  },
  "trafficViolationResponse": {
    "Fine": {
      "Amount": 1000,
      "Points": 7
    },
    "Suspended": "Yes"
  },
  "violation": {
    "Code": null,
    "Date": null,
    "Type": "speed",
    "Speed Limit": 100,
    "Actual Speed": 140
  }
}
----

In the remote method of integrating a process and decision microservices, you can use different tasks to perform the calls to the decision microservices.

.Prerequisites
* You have created a project and have included any Java objects required for your {KOGITO} microservice.
* You have created or imported Business Process Model and Notation (BPMN) 2.0 business processes, Decision Model and Notation (DMN) decision models, Drools Rule Language (DRL) business rules, and other assets in the `src/main/resources` folder of your project.

.Procedure
. Open the new or imported BPMN file to view the process microservice.

> . To map the data between a process microservice and decision microservices, align all the attributes names in a Java class, which is used as process variables.

Reviewer: ```suggestion
. To map data between a process microservice and decision microservices, align all of the attribute names in a Java class, which is used as process variables.
```

Reviewer: I still don't understand what it means to align all the attribute names....

Reviewer: And what is used as process variables? The attribute names?

Author: no the Java class that contains attribute names

Author: Align means, following the consistency

+
--

> If the attributes names contain spaces or not following Java conventions, then you can use https://github.com/FasterXML/jackson[Jackson] annotations. The Jackson annotations align the process variables with the data types of the DMN model.

Reviewer: ```suggestion
If the attributes names contain spaces or do not follow Java conventions, then you can use https://github.com/FasterXML/jackson[Jackson] annotations. The Jackson annotations align the process variables with the data types of the DMN model.
```

Author: Done
